README.md
netlify-git-branch
==================

[![CI](https://github.com/mrloop/netlify-git-branch/actions/workflows/ci.yml/badge.svg)](https://github.com/mrloop/netlify-git-branch/actions/workflows/ci.yml)
[![Latest NPM release][npm-badge]][npm-badge-url]

[npm-badge]: https://img.shields.io/npm/v/netlify-git-branch.svg
[npm-badge-url]: https://www.npmjs.com/package/netlify-git-branch

Feature branch staging environments with netlify and git.

Easily deploy feature branches to indiviudal subdomains on netlify.

<!-- toc -->
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->
# Usage
<!-- usage -->
```sh-session
$ npm install -g netlify-git-branch
$ netlify-git-branch COMMAND
running command...
$ netlify-git-branch (--version)
netlify-git-branch/2.0.0 linux-x64 node-v16.13.0
$ netlify-git-branch --help [COMMAND]
USAGE
  $ netlify-git-branch COMMAND
...
```
<!-- usagestop -->
# Commands
<!-- commands -->
* [`netlify-git-branch delete NAME`](#netlify-git-branch-delete-name)
* [`netlify-git-branch deploy NAME`](#netlify-git-branch-deploy-name)
* [`netlify-git-branch help [COMMAND]`](#netlify-git-branch-help-command)

## `netlify-git-branch delete NAME`

Delete netlify staging subdomain for this branch

```
USAGE
  $ netlify-git-branch delete [NAME]

ARGUMENTS
  NAME  Name is prefixed onto the git branch name. For Example on branch 'my-feature' and a name of 'my-site' the domain
        will be 'https://my-site-my-feature.netlify.app'.

DESCRIPTION
  Delete netlify staging subdomain for this branch

EXAMPLES
  $ netlify-git-branch delete my-site
```

_See code: [src/commands/delete/index.ts](https://github.com/mrloop/netlify-git-branch/blob/v2.0.0/src/commands/delete/index.ts)_

## `netlify-git-branch deploy NAME`

Deploy branch to netlify staging subdomain

```
USAGE
  $ netlify-git-branch deploy [NAME] [--dir <value>] [--assert <value>]

ARGUMENTS
  NAME  Name is prefixed onto the git branch name. For Example on branch 'my-feature' and a name of 'my-site' the domain
        will be 'https://my-site-my-feature.netlify.app'.

FLAGS
  --assert=<value>  Check deployed site using CSS <selector>
  --dir=<value>     [default: dist] Deploy site from <dir>

DESCRIPTION
  Deploy branch to netlify staging subdomain

EXAMPLES
  $ netlify-git-branch deploy my-site --dir dist
```

_See code: [src/commands/deploy/index.ts](https://github.com/mrloop/netlify-git-branch/blob/v2.0.0/src/commands/deploy/index.ts)_

## `netlify-git-branch help [COMMAND]`

display help for netlify-git-branch

```
USAGE
  $ netlify-git-branch help [COMMAND] [--json] [--all]

ARGUMENTS
  COMMAND  command to show help for

FLAGS
  --all   see all commands in CLI
  --json  Format output as json.

DESCRIPTION
  display help for netlify-git-branch
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v4.0.3/src/commands/help.ts)_
<!-- commandsstop -->


CHANGELOG.md



## v2.0.0 (2021-12-02)

#### :rocket: Enhancement
* [#25](https://github.com/mrloop/netlify-git-branch/pull/25) Use Oclif and rename to netlify-git-branch ([@mrloop](https://github.com/mrloop))

#### :bug: Bug Fix
* [#10](https://github.com/mrloop/netlify-git-branch/pull/10) Use correct path for long named bin ([@mrloop](https://github.com/mrloop))

#### :house: Internal
* [#9](https://github.com/mrloop/netlify-git-branch/pull/9) Help avoid hitting 63 limit for domain name ([@mrloop](https://github.com/mrloop))
* [#4](https://github.com/mrloop/netlify-git-branch/pull/4) Configure renovatebot ([@mrloop](https://github.com/mrloop))
* [#3](https://github.com/mrloop/netlify-git-branch/pull/3) Add tests ([@mrloop](https://github.com/mrloop))

#### Committers: 1
- Ewan McDougall ([@mrloop](https://github.com/mrloop))

## v1.0.2 (2021-11-09)

#### :bug: Bug Fix
* [#2](https://github.com/mrloop/netlify-git-branch/pull/2) Proper dependencies ([@mrloop](https://github.com/mrloop))

#### Committers: 1
- Ewan McDougall ([@mrloop](https://github.com/mrloop))

## v1.0.1 (2021-11-09)

#### :house: Internal
* [#1](https://github.com/mrloop/netlify-git-branch/pull/1) Configure release-it ([@mrloop](https://github.com/mrloop))

#### Committers: 1
- Ewan McDougall ([@mrloop](https://github.com/mrloop))

# Changelog


tests/test.js
import { execa } from "execa";
import crypto from "crypto";
import QUnit from "qunit";

let { module, test } = QUnit;

let str = crypto.randomBytes(2).toString("hex");

module("for current branch", function () {
  test("deploy netlify site", async function (assert) {
    let { stdout } = await execa(
      "../bin/run.js",
      [`test-${str}`, ".test-class"],
      {
        env: { DEBUG: true },
        cwd: "tests",
      }
    );
    let regex = new RegExp(`Site not found: test-${str}-.*`);
    assert.true(regex.test(stdout), `${regex.toString()}\n${stdout}`);

    regex = new RegExp(`Site created: http://test-${str}-.*.netlify.app`);
    assert.true(regex.test(stdout), `${regex.toString()}\n${stdout}`);

    regex = new RegExp(`Deployed: http://test-${str}-.*.netlify.app`);
    assert.true(regex.test(stdout), `${regex.toString()}\n${stdout}`);
  });

  test("delete netlify site", async function (assert) {
    let { stdout } = await execa("../bin/run.js", [`test-${str}`, "-d"], {
      env: { DEBUG: true },
      cwd: "tests",
    });
    let regex = new RegExp(`Site deleted: http://test-${str}-.*.netlify.app`);
    assert.true(regex.test(stdout), `${regex.toString()}\n${stdout}`);
  });
});


test/commands/index.test.ts
import {expect, test} from '@oclif/test'
import * as crypto from 'crypto'

const str = crypto.randomBytes(2).toString('hex')
const name = `test-${str}`

describe('commands', () => {
  describe('deploy', () => {
    test
    .stderr({print: true})
    .command(['deploy', name, '--dir', 'test/dist', '--assert', '.test-class'])
    .it('runs deploy cmd', ctx => {
      let regex = new RegExp(`Site not found: test-${str}-.*`)
      expect(ctx.stderr).to.match(regex)
      regex = new RegExp(`Site created: http://test-${str}-.*.netlify.app`)
      expect(ctx.stderr).to.match(regex)
      regex = new RegExp(`Deployed: http://test-${str}-.*.netlify.app`)
      expect(ctx.stderr).to.match(regex)
    })
  })

  describe('delete', () => {
    test
    .stderr({print: true})
    .command(['delete', name])
    .it('runs delete cmd', ctx => {
      expect(ctx.stderr).to.match(new RegExp(`Site deleted: http://test-${str}-.*.netlify.app`))
    })
  })
})


src/index.js
import process from "process";
import puppeteer from "puppeteer";
import { execa } from "execa";
import NetlifyApi from "netlify";
import { deploySite } from "netlify-cli/src/utils/deploy/deploy-site.js";

let netlify = new NetlifyApi(process.env.NETLIFY_AUTH_TOKEN);

function debug(data, ...args) {
  if (process.env.DEBUG) {
    console.log(data, ...args);
  }
}

// https://stackoverflow.com/questions/58033366/how-to-get-current-branch-within-github-actions
// https://docs.github.com/en/actions/reference/environment-variables
async function branchName() {
  let branchName = process.env.GITHUB_HEAD_REF;
  if (!branchName) {
    let { stdout } = await execa("git", ["branch", "--show-current"]);
    branchName = stdout.trim();
  }
  return branchName.replace(/[./]/g, "-");
}

async function revision() {
  let { stdout: revision } = await execa("git", ["log", "--pretty=%h", "-n1"]);
  return revision.trim();
}

export async function siteName(prefix) {
  return `${prefix}-${await branchName()}`;
}

async function findSite(siteName) {
  let sites = await netlify.listSites();
  return sites.find(({ name }) => name === siteName);
}

export async function deleteSite(siteName) {
  let site = await findSite(siteName);
  if (site) {
    await netlify.deleteSite({ site_id: site.id });
    debug("Site deleted:", site.url);
  } else {
    debug("Site not found:", siteName);
  }
}

async function findOrCreateSite(name) {
  let site = await findSite(name);
  if (site) {
    debug("Site found:", name);
  } else {
    debug("Site not found:", name);
    site = await netlify.createSite({ body: { name } });
    debug("Site created:", site.url);
  }
  return site;
}

async function deploy(siteName, folder) {
  let { id } = await findOrCreateSite(siteName);
  let message = `Revision ${await revision()}`;
  debug("Deploying:", message);
  return await deploySite(netlify, id, folder, {
    message,
    filter: () => true,
  });
}

async function checkDeploy(url, selector) {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto(url);
  await page.waitForSelector(selector, { timeout: 10000 });
  await browser.close();
}

export async function deployAndCheck(siteName, cssSelectorToCheck) {
  let { deploy: d } = await deploy(siteName, "dist");
  await checkDeploy(d.url, cssSelectorToCheck);
  debug("Deployed:", d.url);
}

export async function run() {
  let args = process.argv.slice(2);
  try {
    let name = await siteName(args[0]);
    if (args[1] === "-d") {
      deleteSite(name);
    } else {
      deployAndCheck(name, args[1]);
    }
  } catch (error) {
    console.error(error);
    process.exitCode = 1;
  }
}


src/commands/deploy/index.ts
import {Command, Flags} from '@oclif/core'
import Site, {nameDescription} from '../../site'

export default class Deploy extends Command {
  static description = 'Deploy branch to netlify staging subdomain';

  static examples = ['$ netlify-git-branch deploy my-site --dir dist'];

  static args = [
    {name: 'name', description: nameDescription, required: true},
  ];

  static flags = {
    dir: Flags.string({
      description: 'Deploy site from <dir>',
      default: 'dist',
    }),
    assert: Flags.string({
      description: 'Check deployed site using CSS <selector>',
      required: false,
    }),
  };

  async run(): Promise<void> {
    const {args, flags} = await this.parse(Deploy)
    const site = new Site(args.name, this.debug)
    const url = await site.deploy(flags.dir)
    if (flags.assert) {
      await site.checkDeploy(url, flags.assert)
    }
    this.debug('Deployed:', url)
  }
}


src/site.ts
import * as NetlifyApi from 'netlify'
import * as puppeteer from 'puppeteer'
import * as execa from 'execa'
import {deploySite} from 'netlify-cli/src/utils/deploy/deploy-site.js'

// https://stackoverflow.com/questions/58033366/how-to-get-current-branch-within-github-actions
// https://docs.github.com/en/actions/reference/environment-variables
declare type Debug = (...args: any[]) => void;

export const nameDescription = "Name is prefixed onto the git branch name. For Example on branch 'my-feature' and a name of 'my-site' the domain will be 'https://my-site-my-feature.netlify.app'."

export default class Site {
  netlify: any;

  constructor(public prefix: string, public debug: Debug) {
    this.prefix = prefix
    this.debug = debug
    this.netlify = new NetlifyApi(process.env.NETLIFY_AUTH_TOKEN)
  }

  async branchName(): Promise<string> {
    let branchName = process.env.GITHUB_HEAD_REF
    if (!branchName) {
      const {stdout} = await execa('git', ['branch', '--show-current'])
      branchName = stdout.trim()
    }
    return branchName?.replace(/[./]/g, '-') || ''
  }

  async revision(): Promise<string> {
    const {stdout: revision} = await execa('git', [
      'log',
      '--pretty=%h',
      '-n1',
    ])
    return revision.trim()
  }

  async name(): Promise<string> {
    return `${this.prefix}-${await this.branchName()}`
  }

  async findSite(): Promise<any> {
    const siteName = await this.name()
    const sites = await this.netlify.listSites()
    return sites.find(({name}: { name: string }) => name === siteName)
  }

  async findOrCreate(): Promise<any> {
    const siteName = await this.name()
    let site = await this.findSite()
    if (site) {
      this.debug('Site found:', siteName)
    } else {
      this.debug('Site not found:', siteName)
      site = await this.netlify.createSite({body: {name: siteName}})
      this.debug('Site created:', site.url)
    }
    return site
  }

  async deploy(folder: string): Promise<string> {
    const {id} = await this.findOrCreate()
    const message = `Revision ${await this.revision()}`
    this.debug('Deploying:', message)
    const {deploy: d} = await deploySite(this.netlify, id, folder, {
      message,
      filter: () => true,
    })
    return d.url
  }

  async checkDeploy(url: string, selector: string): Promise<void> {
    const browser = await puppeteer.launch()
    const page = await browser.newPage()
    await page.goto(url)
    await page.waitForSelector(selector, {timeout: 10000})
    await browser.close()
  }

  async delete(): Promise<void> {
    const siteName = await this.name()
    const site = await this.findSite()
    if (site) {
      /* eslint-disable camelcase */
      await this.netlify.deleteSite({site_id: site.id})
      this.debug('Site deleted:', site.url)
    } else {
      this.debug('Site not found:', siteName)
    }
  }
}
